src/AudioManager.js

```diff
@@ -85,66 +87,68 @@ export default class AudioManager {
         left: Sampler,
         right: Sampler
     };
+    webAudioApiIsAvailable: boolean;
```

Comment:
My thinking here is that the code will be easier to maintain if we manage the feature detection in the `App` and then select whether to create a real `AudioManager` or a fake one. If we do feature detection at the `AudioManager` level, the checks against `webAudioApiIsAvailable` will be distributed throughout the class. For example, we currently have a check on `playSoundForCharacterState()` but not `playAnnouncement()`. We don't see an error because the `App` test doesn't exercise `playAnnouncement()` but if our `App` test does more in the future, we'd need to make sure that the `AudioManager` public API has sufficient checks against `webAudioApiIsAvailable`.

Reply:
There will be some work setting up types. But an example to follow would be the RobotDriver interface:

https://github.com/codelearncreate/c2lc-coding-environment/blob/37ce5f8c315c44f12d541a1fe104a63e6570879b/src/types.js#L14-L19

Which is implemented by https://github.com/codelearncreate/c2lc-coding-environment/blob/production/src/DashDriver.js and https://github.com/codelearncreate/c2lc-coding-environment/blob/production/src/FakeRobotDriver.js

Reply:
That makes sense, but since switching these is hard coded at the moment for the fake robot driver (and since the speech recognition stuff isn't live either), I don't have a good example of an existing approach.

I assume it's fine to conditionally import based on the feature detection?  I'm not sure how this will affect Flow checking, and thought you might know from previous work with feature detection and mocks.

I'll give the interface and fake implementation approach a try, probably tomorrow.

Reply:
We could conditionally import, but then we need to deal with waiting until the code is loaded. We have an example of doing delayed loading in a component that we are no longer using (also audio based and checks for `AudioContext`) which was part of an early experiment trying speech input:

https://github.com/codelearncreate/c2lc-coding-environment/blob/37ce5f8c315c44f12d541a1fe104a63e6570879b/src/MicMonitor.js#L27-L35

In this case though, I think we can safely go ahead and import both types and select which to instantiate at runtime. We needed to use to conditional import for the component above because it was creating objects at import. Whereas in our Tone.js case, we are seeing the errors about audio objects missing at construction of the `AudioManager`.

I'm thinking something like the following:

```
export default class App extends React.Component<{}, AppState> {
    ...
    audioManager: AudioManager;
    ...
    constructor(props: any) {
        if (webAudioApiIsAvailable()) {
            this.audioManager = new AudioManagerImpl(this.state.audioEnabled);
        } else {
            this.audioManager = new FakeAudioManager();
        }
    }
}
```

Where:

- `AudioManager` is an interface with the public API for the `AudioManager`
- `AudioManagerImpl` is our regular implementation
- `FakeAudioManager` is a minimal fake implementation of the `AudioManager` interface

Reply:
That does seem like a good strategy, I'll make a first pass at that tomorrow.

Reply:
I had just enough time to make a first pass at this.